evap/rewards/tools.py

```diff
@@ -71,15 +73,16 @@ def grant_reward_points(user, semester):
     required_courses = Course.objects.filter(participants=user, semester=semester, is_required_for_reward=True)
     if not required_courses.exists():
         return False
-    # does the user not participate in any more required courses in this semester?
-    if required_courses.exclude(voters=user).exists():
-        return False
-    # did the user not already get reward points for this semester?
-    if RewardPointGranting.objects.filter(user_profile=user, semester=semester).exists():
-        return False
-    # grant reward points
-    RewardPointGranting.objects.create(user_profile=user, semester=semester, value=settings.REWARD_POINTS_PER_SEMESTER)
-    return True
+
+    # How many points have been granted to this user vs how many should they have (this semester)
+    granted_points = RewardPointGranting.objects.filter(user_profile=user, semester=semester).aggregate(Sum('value'))['value__sum'] or 0
+    progress = float(required_courses.filter(voters=user).count()) / float(required_courses.count())
+    target_points = max([points for threshold, points in settings.REWARD_POINTS if threshold <= progress], default=0)
+
+    if target_points > granted_points:
+        RewardPointGranting.objects.create(user_profile=user, semester=semester, value=target_points-granted_points)
+        return True
+    return False
 
 # Signal handlers
 
```

Comment:
the message in `grant_reward_points_after_evaluate` needs to be changed, because `grant_reward_points` can return `True` before all courses have been evaluated.

Reply:
unfortunately this was below the area where i could add comments - it's some lines down from here ;)

Reply:
We could do all kinds of stuff like showing the amount, wether there are more courses to evaluate and so on. But I think keeping it simple is fine, as the user can just check out the rewards page when and if they're interested in the details.

Reply:
"points" is misleading if it's just one ;)

Reply:
but that's probably not too important. however, i would like the text to mention, that more points can be earned for motivational reasons.